janus_core/helpers/correlator.py

```diff
@@ -0,0 +1,412 @@
+        True if the shift registers have data.
+
+        Parameters
+        ----------
+        block : int
+            Block to check the shift register of.
+        p_i : int
+            Index i in the shift (left correland).
+        p_j : int
+            Index j in the shift (right correland).
+
+        Returns
+        -------
+        bool
+            Whether the shift indices have data.
+        """
+        return self._shift_not_null[block, p_i] and self._shift_not_null[block, p_j]
+
+    def get(self) -> tuple[Iterable[float], Iterable[float]]:
+        """
+        Obtain the correlation and lag times.
+
+        Returns
+        -------
+        tuple[Iterable[float], Iterable[float]]
+            The correlation and lags.
+        """
+        correlation = np.zeros(self._points * self._blocks)
+        lags = np.zeros(self._points * self._blocks)
+
+        lag = 0
+        for i in range(self._points):
+            if self._count_correlated[0, i] > 0:
+                correlation[lag] = (
+                    self._correlation[0, i] / self._count_correlated[0, i]
+                )
+                lags[lag] = i
+                lag += 1
+        for k in range(1, self._max_block_used):
+            for i in range(self._min_dist, self._points):
+                if self._count_correlated[k, i] > 0:
+                    correlation[lag] = (
+                        self._correlation[k, i] / self._count_correlated[k, i]
+                    )
+                    lags[lag] = float(i) * float(self._window) ** k
+                    lag += 1
+        return (correlation[0:lag], lags[0:lag])
+
+
+# pylint: disable=R0903
+class Observable:
+    """An abstract observable quantity."""
+
+    @staticmethod
+    def _component_to_index(component: str, *, voigt: bool = False) -> int:
+        """
+        Convert a component code to an index.
+
+        Parameters
+        ----------
+        component : str
+            The component code, x, y, z, ....
+        voigt : bool
+            Use voigt indexing for matrices.
+
+        Returns
+        -------
+        int
+            An index used internally to access the component.
+        """
+        vector = {"x": 0, "y": 1, "z": 2}
+        matrix = {
+            "xx": 0,
+            "xy": 1,
+            "xz": 2,
+            "yx": 3,
+            "yy": 4,
+            "yz": 5,
+            "zx": 6,
+            "zy": 7,
+            "zz": 8,
+        }
+        matrix_voigt = {"xx": 0, "yy": 1, "zz": 2, "yz": 3, "xz": 4, "xy": 5}
+
+        if voigt:
+            return matrix_voigt[component]
+        if component in vector:
+            return vector[component]
+        return matrix[component]
+
+    @abstractmethod
+    # pylint: disable=C0103
+    def get(self, md: Union[Langevin, VelocityVerlet, ASE_NPT]) -> float:
+        """
+        Get an observable component from md.
+
+        Parameters
+        ----------
+        md : Union[Langevin, VelocityVerlet, ASE_NPT]
```

Comment:
```suggestion
    def get(self, md: Union[Langevin, VelocityVerlet, ASE_NPT]) -> float:
        """
        Get an observable component from md.

        Parameters
        ----------
        md : Union[Langevin, VelocityVerlet, ASE_NPT]
```
I think we've used this elsewhere, but maybe it might be neater to use `MolecularDynamics` from `ase.md.md`, which would make it easier to extend? There's nothing inherent preventing this method working with other MD classes.